### PR TITLE
fix: validate compact JSONB number continuations

### LINKED ISSUE #3883: [BUG] Multiple ArrayIndexOutOfBoundsException in JSONReaderUTF8.java, JSONReaderASCII.java, JSONReader.java
### Description

ArrayIndexOutOfBoundsException is thrown when JSON.parse(new String("\0.")); called

### Environment

 - OS：  Ubuntu 22.04 16Core 32 GB
 - JDK： OpenJDK Runtime Environment (build 17.0.13+11-Ubuntu-2ubuntu122.04)
 - commit：67871d7a0c36d2493d5594b57a717cce6d1f6787

### Steps to reproduce:

1. Use method `JSON.parse`
2. Input data - string of any legth, which ends with "\0." where "\0" is null-byte (int('\0') = 0)
3. Error happens
```java
public class TestIssue{
    public static void main(String[] args){
        try{
            JSON.parse(new String("\0."));
        } catch (JSONException ignored) {
            System.out.println("Got JSONException");
        }
    }
}
```

### Correct result
`Got JSONException`

### Actual Result
```
Exception: java.lang.ArrayIndexOutOfBoundsException: Index 2 out of bounds for length 2
	at com.alibaba.fastjson2.JSONReaderUTF8.readNumber0(JSONReaderUTF8.java:5239)
	at com.alibaba.fastjson2.JSONReader.readNumber(JSONReader.java:2239)
	at com.alibaba.fastjson2.reader.ObjectReaderImplObject.readObject(ObjectReaderImplObject.java:290)
	at com.alibaba.fastjson2.JSON.parse(JSON.java:150)
```

#### Additional information

In my opition JSONException must be thrown in this case but not unhandled ArrayIndexOutOfBounds exception

Found with Jazzer

### LINKED ISSUE #7669: [BUG] JSONB BC_BIGINT 声明大小 OOM
### 问题描述

fastjson2 的 JSONB 二进制解码器在处理 `BC_BIGINT` 类型标记时，直接读取一个 32 位有符号整数作为 BigInteger 字节数组的长度，并立即调用 `new byte[len]` 分配内存，**无任何上限校验**。攻击者可通过发送仅 6 字节的 JSONB payload，声明 `Integer.MAX_VALUE`（约 2 GB）的分配大小，导致 JVM 抛出 `OutOfMemoryError`，使目标进程崩溃或不稳定。该漏洞不需要认证，也不需要了解目标服务的 schema。

### 环境信息

 - OS信息：跨平台均可复现（已在 Linux x86_64 上验证）
 - JDK信息：JDK 8 及以上均受影响（已在 OpenJDK 21 上验证）；OOM 在任意默认堆配置下均可触发，无需特殊 `-Xmx` 设置
 - 版本信息：fastjson2 2.0.61（`com.alibaba.fastjson2:fastjson2`）

### 重现步骤

1. 引入 `com.alibaba.fastjson2:fastjson2:2.0.61` 依赖
2. 构造 6 字节 JSONB payload：`0xBB 0x48 0x7F 0xFF 0xFF 0xFF`
3. 调用 `JSONB.parse(payload)`，观察 `OutOfMemoryError` 在毫秒内触发

```java
import com.alibaba.fastjson2.JSONB;

public class Main {
    public static void main(String[] args) {
        // Payload 说明：
        //   0xBB        = BC_BIGINT 类型标记
        //   0x48        = BC_INT32 子类型（后续 4 字节为 32 位整数）
        //   0x7FFFFFFF  = Integer.MAX_VALUE = 2,147,483,647 字节（约 2 GB）
        byte[] payload = {
            (byte) 0xBB,
            (byte) 0x48,
            (byte) 0x7F, (byte) 0xFF, (byte) 0xFF, (byte) 0xFF
        };
        System.out.println("Payload 大小：" + payload.length + " 字节");
        System.out.println("声明分配长度：2,147,483,647 字节（Integer.MAX_VALUE）");
        System.out.println("调用 JSONB.parse() ...");
        long t0 = System.currentTimeMillis();
        try {
            Object result = JSONB.parse(payload);
            System.out.println("未触发 OOM（非预期）：" + result);
        } catch (OutOfMemoryError e) {
            System.out.printf("确认 OOM，耗时 %d ms：%s%n",
                System.currentTimeMillis() - t0, e.getMessage());
        } catch (Exception e) {
            System.out.printf("异常：%s: %s%n",
                e.getClass().getName(), e.getMessage());
        }
    }
}
```

Maven 依赖：

```xml
<dependency>
    <groupId>com.alibaba.fastjson2</groupId>
    <artifactId>fastjson2</artifactId>
    <version>2.0.61</version>
</dependency>
```

<details>
<summary>▶ 一键 Docker 复现（仅需 Docker，无需本地 JDK 或 Maven）</summary>

```bash
#!/bin/bash
# 完全自包含的复现脚本，宿主机只需安装 Docker。
# 预期最后一行：CONFIRMED OOM after ...ms: Requested array size exceeds VM limit

cat > /tmp/fastjson2-jsonb-oom-repro.sh << 'SAVE_EOF'
docker run --rm -i maven:3.9-eclipse-temurin-21 bash -s << 'OUTER_EOF'
set -e
mkdir -p /w && cd /w

cat > Main.java << 'EOF_JAVA'
import com.alibaba.fastjson2.JSONB;

public class Main {
    public static void main(String[] args) {
        // BC_BIGINT(0xBB) + BC_INT32(0x48) + Integer.MAX_VALUE (0x7FFFFFFF = ~2GB)
        // 无需设置 -Xmx：Java 在任意默认 JVM 配置下均无法分配单个 2GB 连续数组
        byte[] payload = {(byte)0xBB, 0x48, 0x7F, (byte)0xFF, (byte)0xFF, (byte)0xFF};
        System.out.println("Payload: " + payload.length + " bytes (BC_BIGINT declares 2,147,483,647 bytes)");
        long t0 = System.currentTimeMillis();
        try {
            JSONB.parse(payload);
            System.out.println("No OOM (unexpected)");
        } catch (OutOfMemoryError e) {
            System.out.println("CONFIRMED OOM after " + (System.currentTimeMillis()-t0) + "ms: " + e.getMessage());
        } catch (Exception e) {
            System.out.println("Exception: " + e.getClass().getName() + ": " + e.getMessage());
        }
    }
}
EOF_JAVA

cat > pom.xml << 'EOF_POM'
<project xmlns="http://maven.apache.org/POM/4.0.0"><modelVersion>4.0.0</modelVersion>
<groupId>poc</groupId><artifactId>poc</artifactId><version>1</version>
<dependencies>
  <dependency>
    <groupId>com.alibaba.fastjson2</groupId>
    <artifactId>fastjson2</artifactId>
    <version>2.0.61</version>
  </dependency>
</dependencies>
</project>
EOF_POM

echo "=== pulling dependencies ==="
mvn -q dependency:copy-dependencies -DoutputDirectory=libs
echo "=== compiling ==="
javac -cp "libs/*" Main.java
echo "=== running ==="
java -cp "libs/*:." Main
OUTER_EOF
SAVE_EOF

bash /tmp/fastjson2-jsonb-oom-repro.sh
```

</details>

### 期待的正确结果

`JSONB.parse()` 在读取 `BC_BIGINT` 长度字段时，应对长度值进行上限校验，超限时抛出 `JSONException`（例如：`BC_BIGINT length out of range: 2147483647`），而不是直接调用 `new byte[len]` 分配攻击者控制的超大内存块。`readLength()` 方法中已有 256 MB 上限检查，此处应当复用该防护。

### 相关日志输出

```
Payload 大小：6 字节
声明分配长度：2,147,483,647 字节（Integer.MAX_VALUE）
调用 JSONB.parse() ...
确认 OOM，耗时 3 ms：Java heap space
```

OOM 在毫秒级内触发；JVM 分配器在实际填充数组前即抛出异常。

#### 附加信息

**根因**：`JSONReaderJSONB.readBigInteger()` 中 `BC_BIGINT` 分支调用 `readInt32Value()` 而非 `readLength()`，绕过了后者在 JSONReaderJSONB.java 第 194 行已有的 256 MB 上限检查（约第 4410 行）：

```java
} else if (type == BC_BIGINT) {
    int len = readInt32Value();     // 读取任意 32 位整数，无上限检查
    byte[] bytes = new byte[len];  // 可分配最多 2 GB —— OOM 在此触发
    System.arraycopy(this.bytes, offset, bytes, 0, len);
    offset += len;
    return new BigInteger(bytes);
}
```

**建议修复（Option A，推荐）**：在 `readBigInteger()` 中添加内联上限检查（建议 4 MB，远高于任何合理的 BigInteger 使用场景）：

```java
} else if (type == BC_BIGINT) {
    int len = readInt32Value();
    if (len < 0 || len > 4 * 1024 * 1024) {
        throw new JSONException("BC_BIGINT length out of range: " + len);
    }
    byte[] bytes = new byte[len];
    // ...
}
```

**建议修复（Option B）**：将 `readInt32Value()` 替换为已有防护的 `readLength()`（256 MB 上限），但仍建议收紧为更小的上限。

**特别影响**：Alibaba Dubbo RPC 生态（JSONB 为其默认二进制序列化格式）。暴露 `BigInteger` 参数且面向不受信任调用方的 Dubbo provider 尤其危险，无需认证即可触发。

**已知相关漏洞**：CVE-2023-25194（Apache Kafka 反序列化 OOM，同类声明大小未校验）、CVE-2021-22569（Protobuf Java OOM，repeated 字段数量未校验）、CVE-2022-45688（fastjson2 深度嵌套栈溢出，同一库的不同 DoS 向量）

### PR DESCRIPTION
**Pull request by DarrenChangJR**
## What this PR does and why

Compact JSONB integer and long encodings require one continuation byte after
their marker. When that byte is missing, `getIntByte` and `getLongByte` read
past the reader's logical end.

This affects more than untyped parsing:

- `38` and `d0` leak `ArrayIndexOutOfBoundsException` through 15 facade and
  direct number-reader paths each;
- compact array/string lengths and integer values nested in containers use the
  same unchecked helpers;
- a one-byte logical slice backed by `{marker, 1}` reads outside the slice and
  returns integer `1`.

This change adds the logical-end check at the two shared helper boundaries.
Malformed inputs now consistently throw `JSONException`, while all callers
retain their existing offset handling. The already guarded three-byte compact
forms and valid encodings are unchanged.

## Tests

- compact int32 and int64 markers across parse, typed facade, any, primitive,
  boxed, number, big-number, float, and double readers;
- compact array and UTF-8 string lengths;
- compact values nested in arrays and objects;
- slices whose backing array contains a byte after the logical end;
- truncated three-byte compact controls;
- valid int32/int64 compact boundary values;
- focused JSONB group: 104 tests, no failures or errors;
- clean full core suite: 7,981 tests, no failures, errors, skips, or dumps.

Exact upstream searches on 2026-08-22 found no active duplicate. Closed #3883
is about malformed text JSON, while closed #7669 / PR #7696 handles declared
`BC_BIGINT` lengths rather than truncated compact-number encodings.

<details>
<summary>中文说明</summary>

## 修改内容和原因

JSONB 的紧凑 int/long 编码在类型 marker 后还需要一个 continuation byte。该字节
缺失时，`getIntByte` 和 `getLongByte` 会越过 reader 的逻辑结束位置读取数据。

影响不只是在无类型解析中抛错：

- `38` 和 `d0` 在各 15 条 facade/direct number-reader 路径中都会泄漏
  `ArrayIndexOutOfBoundsException`；
- array/string length 以及容器中的紧凑整数值共用同一组 helper；
- 如果逻辑长度为 1 的 slice 后面在 backing array 中还有字节 1，reader 会越界读取
  并返回整数 1。

本修改在两个共享 helper 中按 reader 的逻辑 `end` 统一检查边界。畸形输入现在
稳定抛出 `JSONException`，所有调用方原有的 offset 处理保持不变。原本已有边界
保护的三字节紧凑编码和合法输入均不受影响。

回归测试覆盖 int32/int64 marker 的 parse、typed facade、any、primitive、boxed、
number、big-number、float、double 路径，array/string length，array/object 内嵌值，
带有范围外 backing byte 的 slice，三字节截断对照，以及合法紧凑边界值。JSONB
focused group 104 项和 clean full core 7,981 项测试全部通过，无 failure、error、
skip 或 dump。2026-08-22 的精确上游检索没有发现 active duplicate；#3883 和
#7669 / #7696 是不同根因。

</details>

### CODE REVIEW

File: core/src/main/java/com/alibaba/fastjson2/JSONReaderJSONB.java [hunk 1]
```diff
@@ -215,15 +215,21 @@ public int readLength() {
         return type;
     }
 
-    private static int getIntByte(byte[] bytes, int offset, int type) {
+    private int getIntByte(byte[] bytes, int offset, int type) {
+        if (offset >= end) {
+            throw outOfBoundsCheckFromToIndex(offset, end);
```
Comment (wenshao) [inline core/src/main/java/com/alibaba/fastjson2/JSONReaderJSONB.java:220]: **[Critical]** The bounds check this PR adds is bypassed by two inline copies of the same decode in this very class. `readTypeHashCode` (~line 1646) decodes the compact int32-byte type-name length inline — `typelen = ((type - BC_INT32_BYTE_ZERO) << 8) + (bytes[offset + 1] & 0xFF);` — and `readTypeHashCode0` (~lines 1737–1739) does the same for `typeIndex`. Neither goes through the now-checked `getIntByte`, so a truncated JSONB buffer that ends right after the type marker still reads past the logical end: a whole-array reader leaks `ArrayIndexOutOfBoundsException`, and a slice/stream reader whose backing array extends past `end` reads out-of-slice bytes successfully, after which the FNV loop consumes up to 2047 bytes outside the valid region — exactly the hazard this PR's `assertMalformedSlice` test pins for the numeric helpers. Concretely, `JSONB.parse(new byte[]{(byte) 0x92, (byte) 121, (byte) 57})` reaches `readTypeHashCode` and reads `bytes[3]` with `end = 3`, while the sibling path this PR fixed correctly throws `JSONException: offset overflow`.

Verified against the unmodified PR build:

```
JSONB.parse({0x92, 121, 57})      -> java.lang.ArrayIndexOutOfBoundsException: Index 3 out of bounds for length 3
JSONB.parse({0x92, 57})           -> java.lang.ArrayIndexOutOfBoundsException: Index 2 out of bounds for length 2
slice readAny, backing > end      -> FNV loop consumes out-of-slice bytes (AIOOBE Index 10 on a 10-byte backing; silent on a larger one)
sibling fixed path, {0x38} alone  -> JSONException: offset overflow, offset 1, end 1
```

Route both sites through the checked helper:

```java
// readTypeHashCode (~line 1645)
typelen = getIntByte(bytes, offset + 1, type);
offset += 2;

// readTypeHashCode0 (~line 1737)
typeIndex = getIntByte(bytes, offset++, strtype);
```

With this applied in a scratch copy, every input above flips to `JSONException: offset overflow`. If you fold this in, please also add the matching regression test — e.g. `assertThrows(JSONException.class, () -> JSONB.parse(new byte[]{BC_TYPED_ANY, BC_STR_ASCII, (byte) (BC_INT32_BYTE_ZERO + 1)}))` — which is red today (throws AIOOBE) and must go red again if the guard is ever removed.

<details><summary>中文说明</summary>

本 PR 新增的边界检查被同一个类中两处相同解码的内联副本绕过了。`readTypeHashCode`（约第 1646 行）内联解码紧凑 int32-byte 类型名长度——`typelen = ((type - BC_INT32_BYTE_ZERO) << 8) + (bytes[offset + 1] & 0xFF);`——`readTypeHashCode0`（约第 1737–1739 行）对 `typeIndex` 做了同样的事。两者都不经过现在已有检查的 `getIntByte`，因此在类型标记后立即截断的 JSONB 缓冲区仍会越过逻辑结束位置读取：整体数组 reader 会泄漏 `ArrayIndexOutOfBoundsException`；backing 数组超出 `end` 的 slice/stream reader 会成功读到 slice 之外的字节，随后 FNV 循环会在有效区域外消耗最多 2047 字节——正是本 PR 的 `assertMalformedSlice` 测试为数字 helper 钉住的危害。具体地，`JSONB.parse(new byte[]{(byte) 0x92, (byte) 121, (byte) 57})` 会走到 `readTypeHashCode` 并在 `end = 3` 时读取 `bytes[3]`，而本 PR 已修复的姊妹路径会正确抛出 `JSONException: offset overflow`。

在未改动的 PR 构建上实测：

```
JSONB.parse({0x92, 121, 57})      -> java.lang.ArrayIndexOutOfBoundsException: Index 3 out of bounds for length 3
JSONB.parse({0x92, 57})           -> java.lang.ArrayIndexOutOfBoundsException: Index 2 out of bounds for length 2
slice readAny，backing > end      -> FNV 循环消耗 slice 之外的字节（10 字节 backing 上 AIOOBE Index 10；更大的 backing 上静默完成）
已修复的姊妹路径，单独 {0x38}      -> JSONException: offset overflow, offset 1, end 1
```

将两处改为走已检查的 helper：

```java
// readTypeHashCode（约第 1645 行）
typelen = getIntByte(bytes, offset + 1, type);
offset += 2;

// readTypeHashCode0（约第 1737 行）
typeIndex = getIntByte(bytes, offset++, strtype);
```

在 scratch 副本中应用后，上述所有输入都翻转为 `JSONException: offset overflow`。如果一并修复，请补充对应回归测试——例如 `assertThrows(JSONException.class, () -> JSONB.parse(new byte[]{BC_TYPED_ANY, BC_STR_ASCII, (byte) (BC_INT32_BYTE_ZERO + 1)}))`——它当前为红（抛 AIOOBE），且守卫被移除时必须再次变红。

</details>

_— qwen3.8-max via Qwen Code /review (v0.22.0)_

Comment (wenshao) [inline core/src/main/java/com/alibaba/fastjson2/JSONReaderJSONB.java:220]: **[Critical]** The guarantee this PR establishes — truncated compact encodings throw `JSONException` instead of reading past the logical end — is defeated in the same `readObject()` value fast path that calls the now-checked `getIntByte` (~line 458). The sibling compact-number guards there are off by one: `isInt32Short(valueType) && offset + 1 < end` before `getInt3(bytes, offset + 1, …)`, whose 2-byte payload sits at `offset+1..offset+2` (needs `offset + 2 < end`), and `valueType == BC_INT32 && offset + 3 < end` before `getIntBE(bytes, offset + 1)`, whose 4-byte payload sits at `offset+1..offset+4` (needs `offset + 4 < end`). In this method the value type byte is not consumed, so the payloads sit one byte further out than the guards assume. The reads go through `UNSAFE.getShort`/`UNSAFE.getInt` (`IOUtils.getShortBE`/`getIntBE`), which perform no bounds check: a whole-array reader later dies with `ArrayIndexOutOfBoundsException`, and a slice reader silently assembles a value from bytes outside the logical slice. The sibling `readArray` fast path uses the correct convention (~lines 1142/1146). Entry points: `JSONB.parseObject(byte[])` directly, and `JSONB.parse` via nested `BC_OBJECT` values.

Verified against the unmodified PR build:

```
JSONB.parseObject({0xA6, 74, 'a', 68, 1})        -> java.lang.ArrayIndexOutOfBoundsException: Index 6 out of bounds for length 5
JSONB.parseObject({0xA6, 74, 'a', 72, 1, 2, 3})  -> java.lang.ArrayIndexOutOfBoundsException: Index 8 out of bounds for length 7
JSONB.parse, nested object short-trunc           -> java.lang.ArrayIndexOutOfBoundsException: Index 9 out of bounds for length 8
slice reader, short-trunc                        -> returned {"a":383} — byte 0x7F read OUTSIDE the logical slice [0,5), no exception
slice reader, int32-trunc                        -> returned {"a":85} — byte 0x55 read outside [0,7), no exception
control (PR-covered byte variant)                -> JSONException: offset overflow, offset 4, end 4
```

Correct the two guards, mirroring `readArray`:

```java
// readObject() ~line 460
} else if (isInt32Short(valueType) && offset + 2 < end) {

// readObject() ~line 464 — same shape as readArray ~lines 1146-1147
} else if (valueType == BC_INT32) {
    int int32Value = getIntBE(bytes, check3(offset + 1, end));
```

With this applied in a scratch tree, all five failing arms flip to `JSONException`, and valid round-trips (`{"byteVal":300,"intVal":100000,"longShort":300,"small":5}`) pass identically on both arms, so the tightened guards do not over-reject. If you fold this in, please extend `truncatedCompactValuesInContainers` with `assertThrows(JSONException.class, () -> JSONB.parseObject(new byte[]{BC_OBJECT, BC_STR_ASCII_FIX_1, 'a', BC_INT32_SHORT_ZERO, 0}))` and the analogous `BC_INT32`-with-3-payload-bytes case — both are red today and must go red again if either guard regresses.

<details><summary>中文说明</summary>

本 PR 确立的保证——截断的紧凑编码抛出 `JSONException` 而不是越过逻辑结束位置读取——在调用新检查 `getIntByte`（约第 458 行）的同一个 `readObject()` 值快路径中被破坏了。那里相邻紧凑数字分支的守卫差了一个字节：`getInt3(bytes, offset + 1, …)` 前的 `isInt32Short(valueType) && offset + 1 < end`——其 2 字节载荷位于 `offset+1..offset+2`（需要 `offset + 2 < end`）；`getIntBE(bytes, offset + 1)` 前的 `valueType == BC_INT32 && offset + 3 < end`——其 4 字节载荷位于 `offset+1..offset+4`（需要 `offset + 4 < end`）。该方法中值类型字节未被消费，因此载荷比守卫假设的位置再靠后一个字节。这些读取经由 `UNSAFE.getShort`/`UNSAFE.getInt`（`IOUtils.getShortBE`/`getIntBE`）进行，不做任何边界检查：整体数组 reader 稍后会以 `ArrayIndexOutOfBoundsException` 崩溃，而 slice reader 会静默地用逻辑 slice 之外的字节拼出数值。姊妹的 `readArray` 快路径使用了正确的约定（约第 1142/1146 行）。入口：`JSONB.parseObject(byte[])` 直达，以及 `JSONB.parse` 经由嵌套的 `BC_OBJECT` 值。

在未改动的 PR 构建上实测：

```
JSONB.parseObject({0xA6, 74, 'a', 68, 1})        -> java.lang.ArrayIndexOutOfBoundsException: Index 6 out of bounds for length 5
JSONB.parseObject({0xA6, 74, 'a', 72, 1, 2, 3})  -> java.lang.ArrayIndexOutOfBoundsException: Index 8 out of bounds for length 7
JSONB.parse，嵌套 object 截断 short               -> java.lang.ArrayIndexOutOfBoundsException: Index 9 out of bounds for length 8
slice reader，截断 short                          -> 返回 {"a":383}——字节 0x7F 取自逻辑 slice [0,5) 之外，无任何异常
slice reader，截断 int32                          -> 返回 {"a":85}——字节 0x55 取自 [0,7) 之外，无任何异常
对照（本 PR 覆盖的 byte 变体）                    -> JSONException: offset overflow, offset 4, end 4
```

参照 `readArray` 修正这两个守卫：

```java
// readObject() 约第 460 行
} else if (isInt32Short(valueType) && offset + 2 < end) {

// readObject() 约第 464 行——与 readArray 约第 1146-1147 行同形
} else if (valueType == BC_INT32) {
    int int32Value = getIntBE(bytes, check3(offset + 1, end));
```

在 scratch 树中应用后，上述五个失败分支全部翻转为 `JSONException`，且合法往返（`{"byteVal":300,"intVal":100000,"longShort":300,"small":5}`）在两侧行为一致，说明收紧的守卫不会误拒合法输入。如果一并修复，请在 `truncatedCompactValuesInContainers` 中补充 `assertThrows(JSONException.class, () -> JSONB.parseObject(new byte[]{BC_OBJECT, BC_STR_ASCII_FIX_1, 'a', BC_INT32_SHORT_ZERO, 0}))` 及类似的带 3 个载荷字节的 `BC_INT32` 用例——两者当前为红，且任一守卫回退时必须再次变红。

</details>

_— qwen3.8-max via Qwen Code /review (v0.22.0)_

**Review by wenshao — CHANGES_REQUESTED** [review]
Not explored to full depth (tool budget reached): `"agent 3b"`: `无**（所有计划内的核查均已完成）。`.

<details>
<summary>中文说明</summary>

未探索到全部深度（达到工具调用预算）：`"agent 3b"`：`无**（所有计划内的核查均已完成）。`。

</details>

_— qwen3.8-max via Qwen Code /review (v0.22.0)_

File: core/src/main/java/com/alibaba/fastjson2/JSONReaderJSONB.java (hunk 1 above)
Comment (DarrenChangJR) [reply core/src/main/java/com/alibaba/fastjson2/JSONReaderJSONB.java:220]: Fixed in c38a68f0c. Both inline type-metadata paths now delegate to the checked getIntByte helper while preserving their cursor advancement. I added whole-array regressions for the typed-name length and direct type-index paths, plus logical-slice regressions for both. The new tests reproduced ArrayIndexOutOfBoundsException before the fix and now produce JSONException. The focused class passes 8/8 tests, and the full core suite passes 7,986 tests with zero failures or errors.

Comment (DarrenChangJR) [reply core/src/main/java/com/alibaba/fastjson2/JSONReaderJSONB.java:220]: Fixed in c38a68f0c. The compact int32-short fast-path guard now requires both payload bytes, and the full int32 path uses check3(offset + 1, end), matching readArray. I added physical-truncation parseObject cases, logical-slice overread cases for both encodings, and valid object controls. The focused class passes 8/8 tests, and the full core suite passes 7,986 tests with zero failures or errors.
